BSON ObjectId is not compatible with Types.ObjectId

### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported


### Mongoose version

6.11.5

### Node.js version

18.13.0

### MongoDB server version

4.16.0

### Typescript version (if applicable)

4.9.5

### Description

```ts
import { ObjectId } from 'mongodb';
import { Types } from 'mongoose';

const objectId: Types.ObjectId = new ObjectId(); // TS2741: Property '_id' is missing in type 'import("~/node_modules/bson/bson").ObjectID' but required in type 'import("mongoose").Types.ObjectId'. 
```

I see the same issue #12537 , and someone fixed it with #13515 on v7.4.0, but not on v6.x.
Since v6.0.0 produced this bug([commit link](https://github.com/Automattic/mongoose/pull/9855/commits/baf3938b928a42a5dedee6e939d5701d06014a28)), why not fix it on v6.x? It's more difficult to migrate to v7.x.

### Steps to Reproduce

it's in the description above, or see the same bug issue #12537 

### Expected Behavior

Fix it on v6.x and suggestions:

## Suggestion 1: Back to the original definition
Removing the `_id` property may not be a good thing:
<img width="640" alt="image" src="https://github.com/Automattic/mongoose/assets/32633065/33b1ffee-93bc-4271-9687-7c810d843c69">

Source code([code link](https://github.com/Automattic/mongoose/blob/be5b7d50d564a5de07b923e35b1a492b4f86f393/lib/index.js#L261)):
<img width="322" alt="image" src="https://github.com/Automattic/mongoose/assets/32633065/d0740369-fbf8-4095-9602-57aa45fd4309">

The original definition([commit link](https://github.com/Automattic/mongoose/pull/9605/commits/01d24b8baac5c5cc75988c77b8565982081c0adb)) seems more reasonable, and it's compatible with bson.ObjectId:
<img width="210" alt="image" src="https://github.com/Automattic/mongoose/assets/32633065/be4310aa-6f89-4b22-af93-fd1675b8df8f">

## Suggestion 2: Removing the magic `_id` on `Types.ObjectId`
<img width="641" alt="image" src="https://github.com/Automattic/mongoose/assets/32633065/457361af-6e5b-4c32-9b9e-16bf8db82d2b">

That's convenience for people using `populate`, but that's not friendly to people using `Types.ObjectId` and `bson.ObjectId`, too confusing.

There may be another way to make it happen on `populate`, but modifying the base ObjectId, sounds not to be a good idea.

```
npx tsc --strict 13712.ts
13712.ts:4:7 - error TS2739: Type 'ObjectId' is missing the following properties from type 'ObjectId': _id, generationTime

4 const objectId: Types.ObjectId = new ObjectId();
        ~~~~~~~~


Found 1 error in 13712.ts:4
```

--- a/package.json
+++ b/package.json
@@ -50,6 +50,7 @@
     "eslint-plugin-markdown": "^3.0.0",
     "eslint-plugin-mocha-no-only": "1.1.1",
     "express": "^4.18.1",
+    "fs-extra": "~11.1.1",
     "highlight.js": "11.7.0",
     "lodash.isequal": "4.5.0",
     "lodash.isequalwith": "4.4.0",
@@ -64,11 +65,10 @@
     "q": "1.5.1",
     "sinon": "15.0.1",
     "stream-browserify": "3.0.0",
-    "tsd": "0.25.0",
+    "tsd": "^0.25.0",
     "typescript": "4.9.5",
     "uuid": "9.0.0",
-    "webpack": "5.75.0",
-    "fs-extra": "~11.1.1"
+    "webpack": "5.75.0"
   },
   "directories": {
     "lib": "./lib/mongoose"

--- a/package.json
+++ b/package.json
@@ -65,7 +65,7 @@
     "q": "1.5.1",
     "sinon": "15.0.1",
     "stream-browserify": "3.0.0",
-    "tsd": "^0.25.0",
+    "tsd": "0.25.0",
     "typescript": "4.9.5",
     "uuid": "9.0.0",
     "webpack": "5.75.0"

--- a/test/types/mongo.test.ts
+++ b/test/types/mongo.test.ts
@@ -1,0 +1,19 @@
+import * as mongoose from 'mongoose';
+import { expectType } from 'tsd';
+import * as bson from 'bson';
+
+import GridFSBucket = mongoose.mongo.GridFSBucket;
+
+function gh12537() {
+  const schema = new mongoose.Schema({ test: String });
+  const model = mongoose.model('Test', schema);
+
+  const doc = new model({});
+
+  const v = new bson.ObjectId('somehex');
+  expectType<string>(v._id.toHexString());
+
+  doc._id = new bson.ObjectId('somehex');
+}
+
+gh12537();

--- a/test/types/mongo.test.ts
+++ b/test/types/mongo.test.ts
@@ -2,8 +2,6 @@ import * as mongoose from 'mongoose';
 import { expectType } from 'tsd';
 import * as bson from 'bson';
 
-import GridFSBucket = mongoose.mongo.GridFSBucket;
-
 function gh12537() {
   const schema = new mongoose.Schema({ test: String });
   const model = mongoose.model('Test', schema);

--- a/types/augmentations.d.ts
+++ b/types/augmentations.d.ts
@@ -1,0 +1,9 @@
+// this import is required so that types get merged instead of completely overwritten
+import 'bson';
+
+declare module 'bson' {
+  interface ObjectId {
+    /** Mongoose automatically adds a conveniency "_id" getter on the base ObjectId class */
+    _id: this;
+  }
+}

--- a/types/index.d.ts
+++ b/types/index.d.ts
@@ -22,6 +22,7 @@
 /// <reference path="./validation.d.ts" />
 /// <reference path="./inferschematype.d.ts" />
 /// <reference path="./virtuals.d.ts" />
+/// <reference path="./augmentations.d.ts" />
 
 declare class NativeDate extends global.Date { }
 

--- a/types/types.d.ts
+++ b/types/types.d.ts
@@ -80,7 +80,7 @@ declare module 'mongoose' {
     }
 
     class ObjectId extends mongodb.ObjectId {
-      _id: this;
+
     }
 
     class Subdocument<IdType = any> extends Document<IdType> {

